binder/src/main/java/io/grpc/binder/internal/PendingAuthListener.java

```diff
@@ -0,0 +1,126 @@
+
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicReference;
+
+import javax.annotation.Nullable;
+
+/**
+ * A {@link ServerCall.Listener} that can be returned by a {@link io.grpc.ServerInterceptor} to
+ * asynchronously advance the gRPC pending resolving a possibly asynchronous security policy check.
+ */
+final class PendingAuthListener<ReqT, RespT> extends ServerCall.Listener<ReqT> {
+
+  private final ConcurrentLinkedQueue<ListenerConsumer<ReqT>> pendingSteps =
+      new ConcurrentLinkedQueue<>();
+  private final AtomicReference<Listener<ReqT>> delegateRef = new AtomicReference<>(null);
+
+  /**
+   * @param authStatusFuture a ListenableFuture holding the result status of the authorization
+   *                         policy from a {@link io.grpc.binder.SecurityPolicy} or a
+   *                         {@link io.grpc.binder.AsyncSecurityPolicy}. The call only progresses
+   *                         if {@link Status#isOk()} is true.
+   * @param executorPool     a pool that can provide at least one Executor under which the result
+   *                         of {@code authStatusFuture} can be handled, progressing the gRPC
+   *                         stages.
+   * @param call             the 'call' parameter from {@link io.grpc.ServerInterceptor}
+   * @param headers          the 'headers' parameter from {@link io.grpc.ServerInterceptor}
+   * @param next             the 'next' parameter from {@link io.grpc.ServerInterceptor}
+   */
+  PendingAuthListener(
+      ListenableFuture<Status> authStatusFuture,
+      ObjectPool<? extends Executor> executorPool,
+      ServerCall<ReqT, RespT> call,
+      Metadata headers,
+      ServerCallHandler<ReqT, RespT> next) {
+    Executor executor = executorPool.getObject();
+    Futures.addCallback(
+        authStatusFuture,
+        new FutureCallback<Status>() {
+          @Override
+          public void onSuccess(Status authStatus) {
+            if (authStatus.isOk()) {
+              delegateRef.set(next.startCall(call, headers));
+              maybeRunPendingSteps();
+            } else {
+              call.close(authStatus, new Metadata());
+            }
+            executorPool.returnObject(executor);
+          }
+
+          @Override
+          public void onFailure(Throwable t) {
+            call.close(
+                Status.INTERNAL.withCause(t).withDescription("Authorization future failed"),
+                new Metadata());
+            executorPool.returnObject(executor);
+            }
+          }, executor);
+    }
+
+  /**
+   * Runs any enqueued step in this ServerCall listener as long as the authorization check is
+   * complete. Otherwise, no-op and returns immediately.
+   */
+  private void maybeRunPendingSteps() {
+    @Nullable Listener<ReqT> delegate = delegateRef.get();
+    if (delegate == null) {
+      return;
+    }
+
+      ListenerConsumer<ReqT> nextStep;
```

Comment:
Without synchronization, couldn't two threads reach this section at the same time and incorrectly play back `pendingSteps` concurrently/in a different order than they occurred? In particular I'm thinking of what happens if one thread is running FutureCallback.onSuccess() above and another thread is running one of the onXXX() ServerCall.Listener event callbacks.

Reply:
I thought the same thing when thinking over this days later. Ideally, I would push the queue processing logic to a sequential executor, but returning the executor to the ObjectPool is harder if we're using it outside the constructor.

Used `synchronized` for now. Let me know if you find it inappropriate.